litgpt/api.py

```diff
@@ -0,0 +1,148 @@
+from litgpt.prompts import load_prompt_style, has_prompt_style, PromptStyle
+from litgpt.tokenizer import Tokenizer
+from litgpt.utils import (
+    extend_checkpoint_dir,
+    get_default_supported_precision,
+    load_checkpoint,
+    parse_devices
+)
+
+
+class LLM:
+    def __init__(
+        self,
+        model: GPT,
+        tokenizer: Tokenizer,
+        prompt_style: PromptStyle,
+        device_type: str,
+        devices: Union[int, str] = 1,
+        checkpoint_dir: Path = None
+    ) -> None:
+        self.model = model
+        self.preprocessor = Preprocessor(tokenizer)
+        self.device_type = device_type
+        self.devices = devices
+        self.prompt_style = prompt_style
+        self.checkpoint_dir = checkpoint_dir
+
+    @classmethod
+    def load(
+        cls, checkpoint_dir: Path, device_type: str = "cuda", devices: int = 1,
+        quantize: Optional[Any] = None, precision: Optional[Any] = None
+    ) -> "LLM":
+
+        if devices > 1:
+            raise NotImplementedError(
+                "Support for multiple devices is currently not implemented."
+            )
+
+        checkpoint_dir = extend_checkpoint_dir(Path(checkpoint_dir))
+        model, tokenizer, device, devices, prompt_style = cls._setup_model_and_tokenizer(
+            checkpoint_dir=checkpoint_dir,
+            precision=precision,
+            device_type=device_type,
+            devices=devices
+        )
+        return cls(
+            model=model, tokenizer=tokenizer, device_type=device_type, devices=devices,
+            prompt_style=prompt_style, checkpoint_dir=checkpoint_dir,
+        )
+
+    def generate(
+        self,
+        prompt: str,
+        max_new_tokens: int = 50,
+        temperature: float = 1.0,
+        top_k: Optional[int] = None,
+        top_p: float = 1.0,
+        eos_id: Optional[int] = None,
+        return_as_token_ids: bool = False
+    ) -> Tuple[str, torch.Tensor]:
+        prompt = self.prompt_style.apply(prompt)
+        input_ids = self.preprocessor.tokenizer.encode(prompt)
+        prompt_length = input_ids.size(0)
+        max_returned_tokens = prompt_length + max_new_tokens
+
+        output_ids = generate_fn(
+            model=self.model,
+            prompt=input_ids.to(self.model.device),
+            max_returned_tokens=max_returned_tokens,
+            temperature=temperature,
+            top_k=top_k,
+            top_p=top_p,
+            eos_id=self.preprocessor.tokenizer.eos_id
+        )
+
+        if self.devices > 1:
+            raise NotImplementedError(
+                "Support for multiple devices is currently not implemented for `generate`."
+            )
+        for block in self.model.transformer.h:
+            block.attn.kv_cache.reset_parameters()
+
+        if return_as_token_ids:
+            return output_ids
+        else:
+            return self.preprocessor.tokenizer.decode(output_ids)
+
+    @classmethod
+    def _setup_model_and_tokenizer(
```

Comment:
Why to have this a separate function and not in the bode of load classmethod?

Reply:
I was thinking if a user wants to take a quick glance at the `load` method, seeing all the implementation details could be a bit overwhelming. But yeah, it's not necessary, we can consolidate it all into one single method.

Reply:
I like the intent. But don't think that it helps in this case.
I prefer to use comments to highlight sections, e.g. 
# 1. Load config
...
# 2. Load the model
...
But it's my personal preference.

Reply:
Thanks for the feedback. I think what makes most sense is to structure it in a way that makes testing it in small units easy. So, I will think about the restructuring when I write the tests.

Reply:
Great idea.
Commenting should the last resort.